### PR TITLE
Update catalina.xml j2 file, artifacts and integration tests

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
### Introduction

Currently the client ID is considered to be unique across the server and as a result the same client ID cannot be used in two different tenants. However, organization bound client ID is a valid requirement and we should provide this capability. This feature requires enabling tenant qualified URLs and tenanted sessions.

When tenant qualified URLs are enabled,
- Tenant will always be available in the context or thread local. Hence the application resolving logic can be modified to retrieve the application using both the consumer key and tenant ID. 
- The proposed solution is to modify the underline DAO implementation to include the tenant ID in places where it query for apps by giving the client ID. All the DB queries will be modified to include the tenant ID.
- The tenant ID will be retrieved using a utility method added to the “IdentityTenantUtil” class.

when tenant-qualified URLs are disabled,
- The tenant won’t be available in the context or thread local.
- The client ID will be considered unique across the server and firstly the tenant has to be resolved.
- The proposed solution is to perform this at the valve level. We’ll be introducing a new valve and when a request hits that valve, if that request is an oauth request and if tenant-qualified urls are disabled, the client id will be extracted and app tenant will be queried. The valve will only be enabled for non tenant qualified cases. 

### Proposed changes in this pull request
- Add OAuth app tenant resolver valve to catalina xml file.
- Update dependencies.
- Integration tests for client ID tenant unification improvement.

### Related Issue
- https://github.com/wso2/product-is/issues/16370

### Related PRs
- https://github.com/wso2/carbon-identity-framework/pull/4847
- https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2131
- https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/240

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: modules/distribution/src/repository/resources/conf/templates/repository/conf/tomcat/catalina-server.xml.j2 [hunk 1]
```diff
@@ -109,6 +109,11 @@
                 <Valve className="org.wso2.carbon.tomcat.ext.valves.CarbonStuckThreadDetectionValve" threshold="600"/>
                 <Valve className="org.wso2.carbon.tomcat.ext.valves.CompositeValve"/>
 
+                <!-- OAuthAppTenantResolverValve is required to be invoked when tenant qualified urls are disabled. This should be invoked before the AuthenticationValve -->
+                {% if tenant_context.enable_tenant_qualified_urls is sameas false %}
+                <Valve className="org.wso2.carbon.identity.auth.valve.OAuthAppTenantResolverValve"/>
```
Comment (pulasthi7) [inline modules/distribution/src/repository/resources/conf/templates/repository/conf/tomcat/catalina-server.xml.j2:114]: don't we have a config to disable the uniqueness of the client ID. if so, this should be enabled based on both the conditions, right?

Comment (ThaminduDilshan) [reply modules/distribution/src/repository/resources/conf/templates/repository/conf/tomcat/catalina-server.xml.j2:114]: There's no such config. By default client ID tenant uniqueness will be shipped and in both of these cases, underline code expects tenant id to be present when retrieving the client application.

File: pom.xml [hunk 2]
```diff
@@ -2282,7 +2282,7 @@
     <properties>
 
         <!--Carbon Identity Framework Version-->
-        <carbon.identity.framework.version>5.25.342</carbon.identity.framework.version>
+        <carbon.identity.framework.version>5.25.346-SNAPSHOT</carbon.identity.framework.version>
```
Comment (sahandilshan) [inline pom.xml:2285]: `SNAPSHOT` is this intentional?

Comment (ThaminduDilshan) [reply pom.xml:2285]: Need to bump to the correct version after merging the dependent PRs

Comment (ThaminduDilshan) [reply pom.xml:2285]: TODO: Version bump

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java [hunk 3]
```diff
@@ -531,6 +556,33 @@ public HttpResponse sendApprovalPostWithConsent(HttpClient client, String sessio
...
+	 * @param client 				http client.
+	 * @param sessionDataKeyConsent session consent data.
+	 * @param consentClaims 		claims requiring user consent.
+	 * @param tenantDomain 			tenant domain.
+	 * @return http response.
+	 * @throws java.io.IOException IOException.
+	 */
+	public HttpResponse sendApprovalPostWithConsent(HttpClient client, String sessionDataKeyConsent,
+													List<NameValuePair> consentClaims, String tenantDomain)
+			throws IOException {
```
Comment (sahandilshan) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java:571]: Formating seems to be off here

Comment (ThaminduDilshan) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/OAuth2ServiceAbstractIntegrationTest.java:571]: Seems to be due to git UI web view.

Screenshot from 2023-09-27 12-49-27

**Review by sahandilshan — DISMISSED** [review]

File: pom.xml (hunk 2 above)
Comment (ThaminduDilshan) [reply pom.xml]: Done

**Review by chamathns — DISMISSED** [review]

**Comment by ThaminduDilshan** [comment]
PR builder gets stuck in the middle. However tests run successfully in local execution.

Screenshot from 2023-10-03 13-49-21

**Review by DMHP — APPROVED** [review]

**Pull request merged**